File: 백준/Gold/5427. 불/불.cc
#include <iostream>
#include <queue>
#include <algorithm>
#include <string.h>
using namespace std;

typedef pair<int, int> p;

int T;
int R, C, Jy, Jx;
char map[1001][1001];
queue<p> fire;
queue<p> JiHoon;

//BFS
int dp[4][2] = { {1,0}, {0,1}, {-1,0}, {0,-1} };
int totalTime = 0;
int visited[1001][1001] = { 0 };

void Insert_And_Init() {
	cin >> T;
}

void Init() {
	string tmp;

	memset(visited, 0, sizeof(visited));
	while (!fire.empty()) fire.pop();
	while (!JiHoon.empty()) JiHoon.pop();

	cin >> C >> R;

	for (int i = 0; i < R; i++) {
		cin >> tmp;
		for (int j = 0; j < C; j++) {
			map[i][j] = tmp[j];
			if (map[i][j] == '@') {
				JiHoon.push(make_pair(i, j));
				Jx = j; Jy = i;
			}
			else if (map[i][j] == '*') fire.push(make_pair(i, j));
		}
	}
}

bool CanGo(int y, int x) {
	if (y >= 0 && x >= 0 && y < R && x < C && visited[y][x] == 0 && map[y][x] == '.') return true;
	return false;
}

bool CanGo2(int y, int x) {
	if (y >= 0 && x >= 0 && y < R && x < C && (map[y][x] == '.' || map[y][x] == '@')) return true;
	return false;
}

int BFS() {
	visited[Jy][Jx] = 1;

	while (!JiHoon.empty()) {

		//불씨 살아 있는 만큼
		int size = fire.size();
		for (int i = 0; i < size; i++) {
			p cur = fire.front(); fire.pop();
			for (int i = 0; i < 4; i++) {
				int nextY = cur.first + dp[i][0];
				int nextX = cur.second + dp[i][1];

				if (CanGo2(nextY, nextX)) {
					map[nextY][nextX] = '*';
					fire.push(make_pair(nextY, nextX));
				}
			}
		}


		//지훈이 살아 있는 만큼
		size = JiHoon.size();
		for (int i = 0; i < size; i++) {
			p cur = JiHoon.front(); JiHoon.pop();

			//다왔으면 탈출
			if (cur.first == 0 || cur.second == 0 || cur.first == R - 1 || cur.second == C - 1) return visited[cur.first][cur.second];

			for (int i = 0; i < 4; i++) {
				int nextY = cur.first + dp[i][0];
				int nextX = cur.second + dp[i][1];

				if (CanGo(nextY, nextX)) {
					visited[nextY][nextX] = visited[cur.first][cur.second] + 1;
					map[nextY][nextX] = '@';
					JiHoon.push(make_pair(nextY, nextX));
				}
			}
		}
	}
	return 0;
}

void Calculate() {
	for (int i = 0; i < T; i++){
		Init();
		int result = BFS();
		if (result == 0) cout << "IMPOSSIBLE\n";
		else cout << result<<"\n";

	}
	
}

int main() {
	ios_base::sync_with_stdio(false); cin.tie(NULL);

	Insert_And_Init();

	Calculate();
}

File: 백준/Gold/5427. 불/불.kt
import java.util.LinkedList

var T = 0

var w = 0
var h = 0

// 상근이
var sangY = 0
var sangX = 0

// 방향
val dx = intArrayOf(1, -1, 0, 0)
val dy = intArrayOf(0, 0, 1, -1)

fun insertAndInit() {
    T = readln().toInt()
}

fun bfs(fire: LinkedList<Pair<Int, Int>>, map: Array<CharArray>): Int {
    val visited = Array(1001){ BooleanArray(1001) { false } }
    val que = LinkedList<Pair<Int, Int>>()

    visited[sangY][sangX] = true
    que.add(Pair(sangY, sangX))

    var second = 0

    while (que.isNotEmpty()){
        // 불 먼저 옮겨붙기
        val fireSize = fire.size
        for (i in 0 until fireSize){
            val curFire = fire.removeFirst()
            repeat(4){ it ->
                val nextY = curFire.first + dy[it]
                val nextX = curFire.second + dx[it]

                if(CanGoFire(nextY, nextX, map)){
                    map[nextY][nextX] = '*'
                    fire.add(Pair(nextY, nextX))
                }
            }
        }

        // 현재 있는 상근이 다 꺼내
        val size = que.size
        for (i in 0 until size){
            val cur = que.removeFirst()

            // 벽에 도착했으면 탈출
            if (cur.first == 0 || cur.first == h-1 || cur.second == 0 || cur.second == w-1){
                return second + 1
            }

            // 아니면 옮겨
            repeat(4){ it ->
                val nextY = cur.first + dy[it]
                val nextX = cur.second + dx[it]

                if (CanGo(nextY, nextX, map, visited)){
                    visited[nextY][nextX] = true
                    que.add(Pair(nextY, nextX))
                }
            }

        }
        second++
    }

    // 불가능한 경우
    return -1
}

fun CanGo(nextY: Int, nextX: Int, map: Array<CharArray>, visited: Array<BooleanArray>): Boolean {
    return nextY in 0 until h && nextX < w && nextX >= 0 && !visited[nextY][nextX] && map[nextY][nextX] == '.'
}


fun CanGoFire(nextY: Int, nextX: Int, map: Array<CharArray>): Boolean {
    return nextY in 0 until h && nextX < w && nextX >= 0 && (map[nextY][nextX] != '#' && map[nextY][nextX] != '*')
}

fun calculate() {
    repeat(T){
        // 불
        val fire = LinkedList<Pair<Int, Int>>()
        // 지도
        val map = Array(1001){ CharArray(1001) }

        // 정보 입력
        val input = readln().split(" ").map { it.toInt() }
        w = input[0]; h = input[1]

        repeat(h){ j ->
            val line = readln()
            line.forEachIndexed{ idx , value ->
                map[j][idx] = value
                // 불 저장
                if (value == '*'){
                    fire.add(Pair(j, idx))
                }
                // 상근이 저장
                if( value == '@'){
                    sangY = j
                    sangX = idx
                }
            }
        }

        // BFS 실행
        val result = bfs(fire, map)
        if (result == -1){
            println("IMPOSSIBLE")
        } else{
            println(result)
        }


    }
}


fun main() = with(System.`in`.bufferedReader()) {
    insertAndInit()

    calculate()
}